src/app/doc_api.cpp

```diff
@@ -261,17 +261,46 @@ bool DocApi::cropCel(LayerImage* layer,
       newCelBounds.offset(-bounds.origin());
 
       gfx::Point paintPos(newCelBounds.x - newCelPos.x,
-                          newCelBounds.y - newCelPos.y);
+        newCelBounds.y - newCelPos.y);
 
+      const color_t bg = image->pixelFormat() == IMAGE_TILEMAP ?
+                           notile :
+                           m_document->bgColor(layer);
       newCelPos = newCelBounds.origin();
 
-      // Crop the image
+      doc::Grid grid;
+      if (layer->isTilemap()) {
+        const Tileset* tileset = static_cast<LayerTilemap*>(layer)->tileset();
+        grid = tileset->grid();
+
+        const gfx::Point roiOrigin = bounds.origin() - cel->position();
+        const gfx::Rect  roiRect(roiOrigin, bounds.size());
+        const gfx::Size  imageTileSize = grid.canvasToTile(roiRect).size();
+        const gfx::Point newPaintPos = grid.canvasToTile(paintPos);
+        const gfx::Point celPaintPixelPos = grid.tileToCanvas(newPaintPos);
+        const gfx::Point celPosDifference = paintPos - celPaintPixelPos;
+        const gfx::Size  imageSizeInPixel = grid.tilemapSizeToCanvas(imageTileSize);
+
+        if (imageSizeInPixel.w >= newCelBounds.w)
+          newCelPos.x -= celPosDifference.x;
+        else
+          newCelPos.x += celPosDifference.x;
+
+        if (imageSizeInPixel.h >= newCelBounds.h)
+          newCelPos.y -= celPosDifference.y;
+        else
+          newCelPos.y += celPosDifference.y;
+
+        paintPos = newPaintPos;
+        newCelBounds.setSize(imageTileSize);
+      }
+
+      // crop the image
       ImageRef newImage(
         crop_image(image,
                    paintPos.x, paintPos.y,
                    newCelBounds.w, newCelBounds.h,
```

Comment:
Remember: to crop correctly a tilemap image with `crop_image` function, `paintPos.x`, `paintPos.y`, `newCelBounds.w`, `newCelBounds.h` these should be in tile units.

(x, y) `paintPos`  is the top-left corner tile coordinate from the `image` where the crop begins.

(w, h) `newCelBounds` is the size (in tile units) of the part of the `image` to being cropped.

Reply:
When you do this:
```
const Tileset* tileset = static_cast<LayerTilemap*>(layer)->tileset();
grid = tileset->grid();
```
`grid.origin()` will be (0, 0) because the `static_cast<LayerTilemap*>(layer)->tile set()` doesn't return grid origin information, that's why we need to adjust the grid.origin. If you didn't adjust the `grid.origin()` the `canvasToTile` function will return incorrect coordinates.
Look at the following example:  correct `canvasToTile` return values (2x2 pixel individual tile size):
![canvasToTile](https://user-images.githubusercontent.com/32165475/134707887-f458734e-0295-4847-bff3-b8ef64ef118b.gif)